### PR TITLE
Fix DataTips in Web apps...

### LINKED ISSUE #2916: DataTips don't work in MVC5 Web App (Razor cshtml)
1.  Create a new C# ASP.NET Web Application (MVC)
2.  Set a breakpoint in ViewStart.cshtml
3.  F5
4.  Hover mouse over a variable or property

**RESULTS:**
after

**EXPECTED RESULTS:**
before

### PR DESCRIPTION
**Pull request by KevinH-MS**
**Bug number** - Fixes #2916
**Customer scenario** - DataTips do not work at all in Web apps.
**Fix description** - We were previously attempting to match spans in the subject buffer with ```Length == 1```.  This was incorrect, because the input span is always expected to be zero length in GetDataTipText.  I've fixed the check and added an Assert to this effect.
**Testing done** - Ad hoc testing of DataTips and Quick Info in an MVC5 Web App (Razor cshtml)

### CODE REVIEW

**Comment by KevinH-MS** [comment]
Potential reviewers: @balajikris @Pilchie @ManishJayaswal @cston @amcasey

File: src/VisualStudio/Core/Def/Implementation/Venus/VenusCommandFilter.cs [hunk 1]
```diff
@@ -52,14 +53,15 @@ public override int GetDataTipText(TextSpan[] pSpan, out string pbstrText)
...
-            // source spans, take the one that had a length of 1 (since that's how long
-            // our input span was).
-            // If we had no such spans, just return.
+            // We need to map the TextSpan from the DataBuffer to our subject buffer. We'll
+            // only consider spans whose length matches our input span (which we expect to
+            // always be zero). This is to address the case where the position is on a seam
+            // and maps to multiple source spans.
+            // If there is not exactly one matching span, just return.
             var span = WpfTextView.TextViewModel.DataBuffer.CurrentSnapshot.GetSpan(pSpan[0]);
+            Debug.Assert(span.Length == 0);
```
Comment (balajikris) [inline src/VisualStudio/Core/Def/Implementation/Venus/VenusCommandFilter.cs]: Nit: consider adding a failure message

**Comment by balajikris** [comment]
Is it feasible to add an unit test or an integration test to prevent future regression?

Tagging @rchande as well.
It seems that the current code* would have worked at some point in time (before it was broken) and it operated by either finding a span of length 1 and failing if none was found.

*Internal tf history shows CS 1064162 for bug 766220

File: src/VisualStudio/Core/Def/Implementation/Venus/VenusCommandFilter.cs (hunk 1 above)
Comment (KevinH-MS) [inline src/VisualStudio/Core/Def/Implementation/Venus/VenusCommandFilter.cs]: Done

**Comment by KevinH-MS** [comment]
RE: tests - I really would like to add an automated test, but the cost/benefit doesn't seem right to me (given the simplicity of the change).  We don't seem to have *any* unit tests for ```VenusCommandFilter```, and I don't think we have a reliable way to automate DataTips via Apex either.

My hope is that the updated comment and the additional Assert make the intent more clear to future developers and mitigate the risk of future regression.

**Comment by Pilchie** [comment]
:+1:

**Comment by balajikris** [comment]
:shipit:

**Comment by ManishJayaswal** [comment]
Approved 5/26 by ML Shiproom.

Please consider adding unit tests for venus command filter

**Pull request merged**